llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,393 @@
+  std::unordered_set<const Type *> TypesToProcess;
+  Type *DoubleTy = Type::getDoubleTy(M.getContext());
+
+  // 6 is taken from sycl/include/CL/sycl/aspects.hpp
+  static constexpr int AspectFP64 = 6;
+  TypesWithAspects[DoubleTy].insert(AspectFP64);
+
+  TypesToProcess.insert(DoubleTy);
+  for (Type *T : M.getIdentifiedStructTypes())
+    TypesToProcess.insert(T);
+
+  TypesEdgesTy Edges;
+  for (const Type *T : TypesToProcess) {
+    for (const Type *TT : T->subtypes()) {
+      // If TT = %A*** then we want to get TT = %A
+      while (TT->isPointerTy())
+        TT = TT->getPointerElementType();
+
+      // We are not interested in some types. For example, IntTy.
+      if (!TypesToProcess.count(TT))
+        continue;
+
+      Edges[TT].push_back(T);
+    }
+  }
+
+  TypeToAspectsMapTy Result;
+  for (const Type *T : TypesToProcess) {
+    propagateAspectsThroughTypes(Edges, T, TypesWithAspects);
+  }
+}
+
+/// Returns all aspects which might be reached from type @T.
+/// It encompases composite structures and pointers.
+/// NB! This function inserts new records in @Types map for new discovered
+/// types. For the best perfomance pass this map in the next invocations.
+AspectsSetTy getAspectsFromType(const Type *T, TypeToAspectsMapTy &Types) {
+  auto It = Types.find(T);
+  if (It != Types.end())
+    return It->second;
+
+  // This is essential to no get into infinite recursive loops.
+  Types[T] = {};
+  AspectsSetTy Result;
+
+  for (const Type *TT : T->subtypes()) {
+    AspectsSetTy Aspects = getAspectsFromType(TT, Types);
+    Result.insert(Aspects.begin(), Aspects.end());
+  }
+
+  Types[T] = Result;
+  return Result;
+}
+
+/// Returns aspects which might be used in instruction @I.
+/// Function inspects return type and all operand's types.
+/// NB! This function inserts new records in @Types map for new discovered
+/// types. For the best perfomance pass this map in the next invocations.
+AspectsSetTy getAspectsUsedByInstruction(const Instruction &I,
+                                         TypeToAspectsMapTy &Types) {
+  Type *ReturnType = I.getType();
+  AspectsSetTy Result = getAspectsFromType(ReturnType, Types);
+  for (const auto &OperandIt : I.operands()) {
+    Type *T = OperandIt->getType();
+    AspectsSetTy Aspects = getAspectsFromType(T, Types);
+    Result.insert(Aspects.begin(), Aspects.end());
+  }
+
+  return Result;
+}
+
+/// Class for emiting warning messages. Unfortunately, LLVMContext
+/// doesn't contain tools for that.
+class MissedAspectDiagnosticInfo : public DiagnosticInfo {
+  Twine Msg;
+
+public:
+  MissedAspectDiagnosticInfo(Twine DiagMsg,
+                             DiagnosticSeverity Severity = DS_Warning)
+      : DiagnosticInfo(DK_Linker, Severity), Msg(std::move(DiagMsg)) {}
+
+  void print(DiagnosticPrinter &DP) const override { DP << Msg; }
+};
+
+void emitWarning(LLVMContext &C, const StringRef Msg) {
+  C.diagnose(MissedAspectDiagnosticInfo(Msg));
+}
```

Comment:
This one liner is called only in one place. I think that it isn't worth defining a function for that.